mutablesecurity/solutions/implementations/lets_encrypt/code.py

```diff
@@ -1,31 +1,466 @@
     def _install() -> None:
-        pass
+
+        apt.update(
+        sudo=True,
+        name="Updates the apt reporisoties",
+        env={"LC_TIME": "en_US.UTF-8"},
+        cache_time=3600,
+        success_exit_codes=[0, 100],
+        )
+
+        apt.packages(
+        sudo=True,
+        name="Installs the requirements",
+        packages=["python3-certbot-nginx", "curl"],
+        latest=True,
+        )
+
+        files.directory(
+        path="/opt/mutablesecurity/lets_encrypt",
+        present=True,
+        name="Creates the folder that will store Let's Encrypt.",
+        )
+
+        GenerateCertificate().ACT()
+
+            
 
     @staticmethod
     @deploy
-    def _uninstall() -> None:
-        pass
+    def _uninstall(remove_logs: bool = True) -> None:
+
+        RevokeCurrentCertificate().ACT()
+
+        server.shell(
+            sudo=True,
+            name="Purges Let's Encrypt x Certbot",
+            commands=["apt purge -y letsencrypt certbot"],
+        )
+
+        files.directory(
+            name="Removes Let's Encrypt from /etc",
+            path="/etc/letsencrypt",
+            present=False,
+        )
+
+        files.directory(
+            name="Removes Let's Encrypt from /root/.local/share/",
+            path="/root/.local/share/letsencrypt/",
+            present=False,
+        )
+
+        files.directory(
+            name="Removes Certbot from /opt/eff.org/",
+            path="/opt/eff.org/certbot/",
+            present=False,
+        )
+
+        files.directory(
+            name="Removes Let's Encrypt from /var/lib/",
+            path="/var/lib/letsencrypt/",
+            present=False,
+        )
+
+        files.directory(
+            name="Removes Let's Encrypt from /var/log/",
+            path="/var/log/letsencrypt/",
+            present=False,
+        )
+
+        server.shell(
+            sudo=True,
+            name="Removes everything using autoremove",
+            commands=["apt -y update && apt -y autoremove"],
+        )
+
+        server.shell(
+            sudo=True,
+            name="Restarts the Nginx service to apply changes",
+            commands=["systemctl restart nginx"],
+        )
 
     @staticmethod
     @deploy
     def _update() -> None:
-        pass
+
+        try:
+
+            apt.shell(
+                sudo=True,
+                name="First step of selecting packages and updating Certbot",
+                commands=["sudo -s"],
+            )
+
+            apt.shell(
+                sudo=True,
+                name="Second step of selecting packages and updating Certbot",
+                commands=["echo certbot hold | dpkg --set-selections"],
+            )
+
+            apt.shell(
+                sudo=True,
+                name="Third step of selecting packages and updating Certbot",
+                commands=["apt-get update && apt-get upgrade"],
+            )
+        except:
+            raise CertbotAlreadyUpdatedException()
```

Comment:
And one more thing: the try-except mechanism here doesn't help. One pyinfra raises an error in the execution thread, it can be only caught in the main thread (namely when the Leader module executes the operations).

Reply:
I am quite a bit lost. How exactly are we supposed to check if the `UPDATE` has been successfully made, if everything should happen only in the Leader module (the main thread)?

Reply:
Check if the version from the `apt` repository is greater than the installed one. If this is not the case, raise the exception from line `465`.

Reply:
Oh, thank you.

Reply:
Fixed as requested.